roofit/roofitcore/src/RooMinimizer.cxx

```diff
@@ -816,8 +799,61 @@ void RooMinimizer::profileStop()
 }
 
 
+ROOT::Math::IMultiGenFunction* RooMinimizer::getFitterMultiGenFcn() const
+{
+   return fitter()->GetFCN();
+}
+
+
+ROOT::Math::IMultiGenFunction* RooMinimizer::getMultiGenFcn() const
+{
+   if (getFitterMultiGenFcn()) {
+      return getFitterMultiGenFcn();
+   } else {
+      switch (_fcnMode) {
+      case FcnMode::classic: {
+         return static_cast<ROOT::Math::IMultiGenFunction *>(static_cast<RooMinimizerFcn *>(_fcn));
+      }
+      case FcnMode::gradient: {
+         auto thing = dynamic_cast<RooGradMinimizerFcn *>(_fcn);
+         if (thing != nullptr) {
+            return static_cast<ROOT::Math::IMultiGenFunction *>(thing);
+         } else {
+            throw std::logic_error("In RooMinimizer::fitterFcn: Minimizer fcnMode was set to gradient, but fit function type does not match RooGradMinimizerFcn!");
+         }
+      }
+      }
```

Comment:
```suggestion
      return static_cast<RooMinimizerFcn *>(_fcn);
```
Are all these checks here necessary? I think the logic_error is already checked for in the new `RooMinimizer` constructor.

Reply:
I think you are correct about the if on line 819. I probably just put that in to be on the safe side while developing. Either that, or maybe the check wasn't there in the constructor first. Either way, I think you are right there.

The switch, though, is not trivial, hopefully you have a solution. The issue is that `RooMinimizer` only has the abstract pointer to `RooAbsMinimizerFcn`. But that does/can not inherit from `IMultiGenFunction` (see other comment https://github.com/root-project/root/pull/8596/files/c294d6fd22aaecf26f9ea6c7b654a3dbbe4eaa64#r669354197). To cast to `IMultiGenFunction`, we must first cast to one of the inherited classes (that itself inherits from `IMultiGenFunction` next to inheriting from `RooAbsMinimizerFcn`) and then cast down to `IMultiGenFunction`.

If you can think of a more elegant way to solve this, I'm more than happy to hear it!

Reply:
Thanks for the explanation, I see now why the inheritance hierachy is like it is.

Reply:
Can you just `dynamic_cast`? That's what you have to do when you need to cross-cast across inheritance chains.